### ISSUE TITLE
Align V-A score scale

### ISSUE THREAD

**Issue by kevinmanuellee** [keimiii/emo-rec#4]
- Understand why EmoNet, DEAM, FindingEmo uses different scale of V-A score
- Propose a plan to align V-A score
- Identify what parts of the workflow will be affected

**Comment by DesmondChoy** [comment]
# Definitions

- Valence: **Mood (pleasantness)**. Left is unpleasant/negative, right is pleasant/positive.
- Arousal: **Energy level**. Low is calm/relaxed, high is excited/tense.

```
            Arousal (energy)
                  ↑
   Tense/angry    |   Excited/joyful
   (low V, high A)|   (high V, high A)
------------------+------------------→ Valence (pleasantness)
   Sad/flat       |   Calm/content
   (low V, low A) |   (high V, low A)
```

The circumplex model of emotion was developed by James Russell. This model suggests that emotions are distributed in a two-dimensional circular space, containing arousal and valence dimensions. 

> “Used to measure how pleasant the room feels (valence) and how energized it is (arousal). Those two dials guide us to music with the same emotional ‘coordinates,’ so the soundtrack always fits the moment.”

# Why the scales differ

## EmoNet

- [-1, 1] for both V and A
- Inherits its scale from AffectNet, the dataset it was trained on
- This normalized range was chosen by the computer vision community for computational convenience - it's symmetric, centered at zero, and works well with neural network activations and loss functions.

## DEAM

- Static: [1, 9], Dynamic: [-10, 10]
- Static annotations use the traditional 9-point Self-Assessment Manikin (SAM) scale, a standard in psychology research.

## FindingEmo

- V: [-3, 3], A: [0, 6]
- Represents a unique design choice for scene-level emotion recognition
- The asymmetric scales reflect the theoretical understanding that arousal is inherently unipolar (you can't be less than "not aroused"), while valence is bipolar (positive vs negative emotions).

# Alignment Strategy

Use a normalized [-1, 1] space as the common reference for both valence and arousal. This choice offers several advantages:
- Computational efficiency: Works naturally with neural network activations
- Semantic clarity: Zero represents true neutrality
- Mathematical simplicity: Linear transformations are straightforward
- Industry alignment: Matches the computer vision standard (AffectNet/EmoNet)

## `EmotionScaleAligner` Class with Defined Mapping Functions

```python

import numpy as np
from typing import Tuple, Union

ArrayLike = Union[float, int, np.ndarray]

class EmotionScaleAligner:
    """
    Unified scale alignment for emotion recognition systems.
    Reference space: valence, arousal in [-1, 1].
    """

    def __init__(self, strict: bool = False, dtype=np.float32):
        self.strict = strict
        self.dtype = dtype

    @staticmethod
    def _check_range(x, lo, hi, name, strict: bool):
        if strict:
            if np.any(x < lo) or np.any(x > hi):
                raise ValueError(f"{name} out of range [{lo}, {hi}].")
        return x

    @staticmethod
    def _clip(x, lo, hi):
        return np.clip(x, lo, hi)

    def _as_array(self, *xs):
        return [np.asarray(x, dtype=self.dtype) for x in xs]

    # ---------- FindingEmo ----------
    def findingemo_to_reference(self, v_fe: ArrayLike, a_fe: ArrayLike) -> Tuple[np.ndarray, np.ndarray]:
        """
        FE: valence in [-3, 3], arousal ('Intensity') in [0, 6] → reference [-1, 1].
        """
        v_fe, a_fe = self._as_array(v_fe, a_fe)
        v_fe = self._check_range(v_fe, -3.0, 3.0, "FE valence", self.strict)
        a_fe = self._check_range(a_fe,  0.0, 6.0, "FE arousal",  self.strict)

        # map, then clip to be safe
        v_ref = self._clip(v_fe / 3.0, -1.0, 1.0)
        a_ref = self._clip((a_fe / 3.0) - 1.0, -1.0, 1.0)
        return v_ref, a_ref

    def reference_to_findingemo(self, v_ref: ArrayLike, a_ref: ArrayLike) -> Tuple[np.ndarray, np.ndarray]:
        """
        reference [-1, 1] → FE: valence [-3, 3], arousal [0, 6].
        """
        v_ref, a_ref = self._as_array(v_ref, a_ref)
        v_ref = self._check_range(v_ref, -1.0, 1.0, "Ref valence", self.strict)
        a_ref = self._check_range(a_ref, -1.0, 1.0, "Ref arousal", self.strict)

        v_fe = self._clip(v_ref * 3.0, -3.0, 3.0)
        a_fe = self._clip((a_ref + 1.0) * 3.0, 0.0, 6.0)
        return v_fe, a_fe

    # ---------- DEAM (Static SAM 1..9) ----------
    def deam_static_to_reference(self, v_deam: ArrayLike, a_deam: ArrayLike) -> Tuple[np.ndarray, np.ndarray]:
        """
        DEAM static SAM [1, 9] → reference [-1, 1].
        """
        v_deam, a_deam = self._as_array(v_deam, a_deam)
        v_deam = self._check_range(v_deam, 1.0, 9.0, "DEAM valence", self.strict)
        a_deam = self._check_range(a_deam, 1.0, 9.0, "DEAM arousal", self.strict)

        v_ref = self._clip((v_deam - 5.0) / 4.0, -1.0, 1.0)
        a_ref = self._clip((a_deam - 5.0) / 4.0, -1.0, 1.0)
        return v_ref, a_ref

    def reference_to_deam_static(self, v_ref: ArrayLike, a_ref: ArrayLike) -> Tuple[np.ndarray, np.ndarray]:
        """
        reference [-1, 1] → DEAM static SAM [1, 9].
        """
        v_ref, a_ref = self._as_array(v_ref, a_ref)
        v_ref = self._check_range(v_ref, -1.0, 1.0, "Ref valence", self.strict)
        a_ref = self._check_range(a_ref, -1.0, 1.0, "Ref arousal", self.strict)

        v_deam = self._clip(5.0 + 4.0 * v_ref, 1.0, 9.0)
        a_deam = self._clip(5.0 + 4.0 * a_ref, 1.0, 9.0)
        return v_deam, a_deam

    # ---------- EmoNet ----------
    def emonet_to_reference(self, v_emonet: ArrayLike, a_emonet: ArrayLike) -> Tuple[np.ndarray, np.ndarray]:
        """
        EmoNet outputs are already in [-1, 1]. We just ensure range and dtype.
        """
        v_emonet, a_emonet = self._as_array(v_emonet, a_emonet)
        v_ref = self._clip(v_emonet, -1.0, 1.0)
        a_ref = self._clip(a_emonet, -1.0, 1.0)
        return v_ref, a_ref

```

@keimiii @jrlyy @kevinmanuellee - for your review. Does it make sense? Is there a better implementation?

**Comment by jrlyy** [comment]
I'm good with this approach because I don't have a better suggestion now. 

So just for understanding if we use -1 to 1 for arousal/intensity
-1 = not intense at all
0 = mildly intense 
1 = very intense

**Comment by DesmondChoy** [comment]
> I'm good with this approach because I don't have a better suggestion now.
> 
> So just for understanding if we use -1 to 1 for arousal/intensity -1 = not intense at all 0 = mildly intense 1 = very intense

Yep correct.
